lib/ocpp/v201/smart_charging.cpp

```diff
@@ -315,12 +316,32 @@ SmartChargingHandler::validate_profile_schedules(ChargingProfile& profile,
     return ProfileValidationResultEnum::Valid;
 }
 
-void SmartChargingHandler::add_profile(int32_t evse_id, ChargingProfile& profile) {
-    if (STATION_WIDE_ID == evse_id) {
-        station_wide_charging_profiles.push_back(profile);
+SetChargingProfileResponse SmartChargingHandler::add_profile(int32_t evse_id, ChargingProfile& profile) {
+    SetChargingProfileResponse response;
+    response.status = ChargingProfileStatusEnum::Accepted;
+
+    auto& profile_storage = (STATION_WIDE_ID == evse_id) ? station_wide_charging_profiles : charging_profiles[evse_id];
+    auto profile_with_id =
+        std::find_if(profile_storage.begin(), profile_storage.end(),
+                     [&profile](const ChargingProfile existing_profile) { return profile.id == existing_profile.id; });
+
+    // K01.FR05 - replace non-ChargingStationExternalConstraints profiles if id exists.
+    if (profile_with_id != profile_storage.end() &&
+        profile_with_id->chargingProfilePurpose != ChargingProfilePurposeEnum::ChargingStationExternalConstraints) {
+        *profile_with_id = profile;
     } else {
-        charging_profiles[evse_id].push_back(profile);
+        profile_storage.push_back(profile);
+    }
+
+    return response;
+}
+
+std::vector<ChargingProfile> SmartChargingHandler::get_profiles() {
```

Comment:
I see that this is used for testing so can I assume this is temporary? I don't see a other reason to expose all the profiles to the outside since this class is responsible for the implementation.

Reply:
I agree that testing behavior is ideal rather than testing implementation. At the moment, we have a way to confirm that adding profiles works as expected. If, in the future, there is a better way to test the behavior without having this interface, we could always revisit, of course.

Reply:
Yep understood, no need to block this for now but I think it's good to look at this again in the future.